### DISCUSSION TITLE
Logging request body during log phase and memory requirement

### DISCUSSION THREAD

**Discussion by Sachin-M-Desai** [Kong/kong#11353] [Help]
Hi,

We have a requirement to log request body to an external endpoint. The logging is performed in the log phase, but this requires that the body be saved in the plugin context in the access handler.

Presently, kong.request.get_raw_body() does the job, but since the body has been saved in the access, it can only get freed on the completion of log phase. 

This demands a lot of runtime memory to temporarily hold the request body. Situation worsens if request bodies are large.

Is there way to get the request body during the logging phase?

Thanks!

**Comment by hanshuebner** [comment]
Hello @Sachin-M-Desai,

if you cannot allocate enough memory to buffer request bodies, you will probably want to buffer them in the file system.  This actually happens automatically when you call `get_raw_body` and the body is larger than [nginx_http_client_body_buffer_size](https://docs.konghq.com/gateway/latest/reference/configuration/#nginx_http_client_body_buffer_size).  You can access the file name that nginx used to buffer the body using [ngx.req.get_body_file](https://github.com/openresty/lua-nginx-module#ngxreqget_body_file).

Hope this helps,
Hans

**Comment by Sachin-M-Desai** [comment]
Hi Hans,

May be i was brief in my query. We are running a plugin in an already deployment system and hence we dont have the liberty to change the buffer related configuration. 

The problem i see here is that a plugin has to make a "copy" of the request payload and hold onto it. There already exists a copy of payload in the system. We are affectively having 2 copies. Wouldnt it be better if the payload was available via get_raw_body even in the logging phase? 

As per the get_body_file part, we are trying to avoid that as we cannot change buffer configs and also, there will be a I/O performance penalty.

**Reply by hanshuebner** [reply]
> The problem i see here is that a plugin has to make a "copy" of the request payload and hold onto it. There already exists a copy of payload in the system. We are affectively having 2 copies. Wouldnt it be better if the payload was available via get_raw_body even in the logging phase?

Why do you need to make a copy of the body?  Do you observe that when you assign the result of calling `get_raw_body` to a variable in the context, that variable does not hold on to the body until your `log` callback is invoked?

**Reply by Sachin-M-Desai** [reply]
We perform request body logging from the log phase and since kong.request.get_raw_body() doesnt work in logging mode we are forced save the request body.

The request body variable if assigned in access phase hold the body all right.

**Reply by hanshuebner** [reply]
Can you assign the body to, say, `kong.ctx.plugin.request_body` in the `access` phase and then access it from there in the `log` phase?

**Reply by Sachin-M-Desai** [reply]
> Can you assign the body to, say, `kong.ctx.plugin.request_body` in the `access` phase and then access it from there in the `log` phase?

This is the design that increases the memory requirement since a copy of the payload needs to be maintained by plugins till logging phase. This is the problem and trying to highlight here

**Reply by hanshuebner** [reply]
If you don't want to hold on to the request body for the whole time of the request, you could send it to your log server in a separate request in the `access` phase.  If you need to log the body together with information received from the upstream in one log request, I see no alternative to storing the body somewhere while it is being processed by the upstream.

What alternative design would remove that requirement?  Do you have a proposal that would change things in that regard?

**Reply by Sachin-M-Desai** [reply]
I feel, it makes plugin much simpler if log phase can provide request body as well via get_raw_body(). Kong doesn't have to buffer it by default but can be collected on request from the table as it is today. Currently we have gone down the path of sending the payload in access phase but it just works around the problem and affects latency.

**Reply by hanshuebner** [reply]
It does not matter whether Kong buffers the request body until the log phase or if your plugin does it. The memory usage will be the same, only that if Kong would do it, we'd have to make sure that the request body is only buffered for those plugins that need it. It does not seem to me that there is enough general usefulness for such a feature at this point, given that the solution (buffering the body in the plugin context) is straightforward and does not seem to have significant drawbacks.
